src/Compilers/CSharp/Portable/Binder/SubsumptionDiagnosticBuilder.cs

```diff
@@ -0,0 +1,323 @@
+﻿// Copyright(c) Microsoft.All Rights Reserved.Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+
+    /// <summary>
+    /// Helper class for binding the pattern switch statement. It helps compute which labels
+    /// are subsumed and/or reachable.
+    /// </summary>
+    internal class SubsumptionDiagnosticBuilder : DecisionTreeBuilder
+    {
+        private readonly DecisionTree _subsumptionTree;
+
+        internal SubsumptionDiagnosticBuilder(Symbol enclosingSymbol,
+                                               Conversions conversions,
+                                               BoundExpression expression)
+            : base(enclosingSymbol, conversions)
+        {
+            _subsumptionTree = DecisionTree.Create(expression, expression.Type, enclosingSymbol);
+        }
+
+        /// <summary>
+        /// Add the case label to the subsumption tree. Return true if the label is reachable
+        /// given the expression and previously added labels.
+        /// </summary>
+        internal bool AddLabel(BoundPatternSwitchLabel label, DiagnosticBag diagnostics, out bool valueMatched)
+        {
+            // Use site diagnostics are reported (and cleared) by this method.
+            // So they should be empty when we start.
+            Debug.Assert(UseSiteDiagnostics.Count == 0);
+
+            valueMatched = false;
+
+            if (label.Syntax.Kind() == SyntaxKind.DefaultSwitchLabel)
+            {
+                // the default case label is always considered reachable.
```

Comment:
`valueMatched` means that the value of the input matched the value in the case label.